Display the help text in the TextareaArray setting

kienstra opened the pull request:
(Testing instructions still valid as of May 5th)

## Before
<img width="554" alt="before-no-help" src="https://user-images.githubusercontent.com/4063887/106958725-5b562d00-66ff-11eb-927b-c13ef5439dcc.png">


## After

<img width="555" alt="multiselect-field" src="https://user-images.githubusercontent.com/4063887/106958740-5f824a80-66ff-11eb-8fb7-c60944495ac9.png">

#### Changes
* Adds help text to `<textarea>` settings in Multiselect, Select, and Radio fields 

#### Testing instructions
1. `composer install && npm install && npm run dev`
2. `/wp-admin` > Custom Blocks > Add New
3. Add a new field
4. Change the Field Type to Select
5. Expected: the new help text displays: <img width="572" alt="help-text" src="https://user-images.githubusercontent.com/4063887/106959895-f26fb480-6700-11eb-8e9d-da10b1ae66ed.png">
6. Change the Field Type to Multiselect
7. Expected: there's new help text under 2 settings: <img width="555" alt="multiselect-field" src="https://user-images.githubusercontent.com/4063887/106959973-103d1980-6701-11eb-93f8-49079f90909d.png">
8. Change the Field Type to Radio
9. Expected: there's new help text:
 <img width="549" alt="radio-new-help-text" src="https://user-images.githubusercontent.com/4063887/106960123-3a8ed700-6701-11eb-8669-8751abfb340f.png">

kienstra committed:
Display the help text for the TextareaArray setting

Without this,
it's really hard to know what to enter.
This existed before,
but it usually had markup.
So instead of doing
__dangerouslySetInnerHTML={},
enter the markup in the React file,
and use other help text if it exists.

kienstra committed:
Let e2e tests run by changing to wp-env

It looks like it
has to be wp-env in CircleCI.

kienstra commented on js/src/edit-block/components/settings/textarea-array.js:
```diff
@@ -40,6 +46,28 @@ const TextareaArray = ( { handleOnChange, setting, value } ) => {
+								'genesis-custom-blocks'
+							) }
+						</p>
+						<p className={ helpClass }>
```
Before, this help text was in `setting.help`, but it has markup.

So instead of doing `dangerouslySetInnerHTML={ setting.default }`, this has the markup in the React file.

kienstra commented on php/Blocks/Controls/Multiselect.php:
```diff
@@ -53,13 +53,7 @@ public function register_settings() {
-					_x( 'foo : Foo', 'Format for the menu values. option_value : Option Name', 'genesis-custom-blocks' ),
-					_x( 'bar : Bar', 'Format for the menu values. option_value : Option Name', 'genesis-custom-blocks' )
-				),
+				'help'     => '',
```
This is now in the [React file](https://github.com/studiopress/genesis-custom-blocks/pull/62/files#r570563346). 

This is `''` here in PHP to not override the React's component's help markup.

kienstra commented on php/Blocks/Controls/Multiselect.php:
```diff
@@ -69,7 +63,7 @@ public function register_settings() {
 				'type'     => 'textarea_array',
 				'default'  => '',
-				'help'     => __( 'Enter each default value on a new line.', 'genesis-custom-blocks' ),
+				'help'     => __( 'Each default value on a new line.', 'genesis-custom-blocks' ),
```
To be similar to the other text. I removed this word from the other text to display on 2 lines instead of 3.

kienstra commented on js/src/edit-block/components/editor.js:
```diff
@@ -74,6 +74,16 @@ import {
+ * @property {string} help A help value that display in the GCB editor.
+ * @property {string} type The setting type, like 'width' or 'text', not a data type like boolean.
+ * @property {*} default The default value.
+ */
```
Not related to this PR, but noticed I forgot to give `Setting` a `@typedef` before.

kienstra commented:
Hi @dreamwhisper,
Thanks for bringing this up, it's tough to know what to enter in the `<textarea>` without help text.

How does this look?

kienstra commented:
Hi @RobStino,
Could you please review this? Thanks!

dreamwhisper commented:
@kienstra I'm having some trouble with the multiselect. 

I see this when the block is on the page. Also, no default appears selected.
- Add the block to the page.
- click outside the block or save, then leave the page and come back.
![Screen Shot 2021-03-09 at 10 25 38 AM](https://user-images.githubusercontent.com/1271053/110495858-ec377400-80c2-11eb-8fd0-d80af0c93be6.png)

kienstra commented:
Hi @dreamwhisper,
Thanks a lot for testing this and raising the issue with the multiselect. I'll look at that.

kienstra commented:
Hi @dreamwhisper,
Thanks, I saw the same thing you saw:

<img width="796" alt="multiselct-block" src="https://user-images.githubusercontent.com/4063887/110526273-3298cd00-80db-11eb-8c35-a33826480f43.png">

I'll work on this when GCB tickets get slotted into a sprint 😄

kienstra committed:
Merge in develop, resolve conflicts.

There were conflicts in:
js/src/edit-block/components/settings/checkbox.js
package.json
php/Blocks/Controls/Multiselect.php
php/Blocks/Controls/Radio.php
php/Blocks/Controls/Select.php
tests/php/Unit/Blocks/Controls/TestMultiselect.php
tests/php/Unit/Blocks/Controls/TestRadio.php
tests/php/Unit/Blocks/Controls/TestSelect.php

In almost all cases, resolve conflicts in favor of this
branch: add/choices-description.
Expect in package.json.

kienstra committed:
Remove sanitize and validate values from settings

ControlSetting doesn't have a ->sanitize or ->validate
property anymore.
So no need to pass that.
These settings are saved in React,
and are not validated.

kienstra committed:
If a field is deleted, also delete it from the previewAttributes

If the field no longer exists,
there's no need to add it to the
block's attributes.

kienstra committed:
See if upgrading server-side-render is possible

There's a new package
that was just released 3 days ago.

kienstra committed:
Revert "See if upgrading server-side-render is possible"

This reverts commit a344c514e1d1094676859f5744ad64cbd3b5fa05.

kienstra committed:
When changing a field name, change the name in the previewAttributes

Without this,
<ServerSideRender> could have a network error.

kienstra committed:
On changing the field type, delete the previewAttribute

The attribute could be of a different type.
For example, changing from a Url field
to Image, the type will change from from
string to integer.
So this prevents an error in ServerSideRender.

kienstra committed:
Refactor preview into <FrontEndPreview> component

kienstra committed:
Possible approach to handling <ServerSideRender> error

kienstra committed:
Simpler approach to front-end preview (if it works)

The server-side will only have the saved post values.
So pass the saved previewAttributes,
not just the edited ones.
This looks to prevent errors in <ServerSideRender>.

kienstra commented on js/src/edit-block/components/header.js:
```diff
@@ -109,10 +109,7 @@ const Header = ( { editorMode, setEditorMode } ) => {
 				<span className="mr-3 text-sm">
-					<PostSavedState
-						forceIsDirty={ false }
-						forceIsSaving={ false }
```
These are essentially `false` by default. So no need to pass `false`.

kienstra committed:
Handle the Front-end preview for new post

If there is nothing saved,
there will be nothing to preview.

kienstra committed:
If there is no field to preview, suggest that they add a field

Maybe if this were more effortless,
it would allow them to click a <button> to do that.

kienstra committed:
Change text to 'edit the block,' as it doesn't need fields

Some blocks won't have fields,
just a template.

kienstra commented on js/src/edit-block/components/settings/textarea-array.js:
```diff
@@ -40,6 +46,28 @@ const TextareaArray = ( { handleOnChange, setting, value } ) => {
+							foo : Foo
+						</p>
+						<p className={ helpClass }>
+							bar : Bar
```
Intentionally not translated

kienstra committed:
See if userEven helps the test pass

kienstra committed:
Fix failed e2e test

It looks like this was a real failure:
one of the issues was a deletion of
something that didn't exist.

kienstra committed:
Make e2e test more robust: query for role, not selector

This will look for a combobox (select)
with a lable, as that's more like

what a user would look for.

kienstra committed:
Add artifacts/ to .gitignore, it looks like the new wordpress/scripts adds that dir

When e2e tests fail,
it looks like it uses that to add a screenshot.

kienstra commented on js/src/edit-block/components/editor-preview.js:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @return {React.ReactElement} The editor preview.
+ */
+const EditorPreview = () => {
```
This file is mainly extracted out of `editor.js`

kienstra commented on js/src/edit-block/components/front-end-preview.js:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @return {React.ReactElement} The front-end preview.
+ */
+const FrontEndPreview = () => {
```
This file is mainly extracted from `editor.js`, with some conditionals in case there's nothing to preview.

kienstra commented on php/Blocks/Controls/Radio.php:
```diff
@@ -46,13 +46,7 @@ public function register_settings() {
-					_x( 'foo : Foo', 'Format for the menu values. option_value : Option Name', 'genesis-custom-blocks' ),
-					_x( 'bar : Bar', 'Format for the menu values. option_value : Option Name', 'genesis-custom-blocks' )
-				),
+				'help'    => '',
```
This is now rendered in React.

kienstra committed:
Pluralize the attributes names and values

newAttributes is actually a
key/value pair, I think.

kienstra committed:
Add a new component for the Multiselect defaults

These should be saved as a string[],
not as an array of objects, like before.

kienstra committed:
Fix the failed PHPUnit test

kienstra commented on js/src/edit-block/components/settings/textarea-default.js:
```diff
@@ -0,0 +1,69 @@
+ * @param {TextareaDefaultProps} props The component props.
+ * @return {React.ReactElement} The textarea default component.
+ */
+const TextareaDefault = ( { handleOnChange, setting, value } ) => {
```
This component should fix the problem with the Multiselect error. It saves the default value as an array of strings, instead of the array of objects it did before.

kienstra commented:
Hi @dreamwhisper,
Thanks a lot for [testing this before](https://github.com/studiopress/genesis-custom-blocks/pull/62#issuecomment-794098288). 

Could you please test it again? The multiselect issue looks to be fixed in my local.

Thanks, Jen!

dreamwhisper reviewed (approved):
Looks good!

kienstra commented:
Thanks so much, Jen!!!
